### ISSUE TITLE
Query Regarding Gradient Clipping in Mamba Implementation

### ISSUE THREAD

**Issue by EricLina** [state-spaces/mamba#291]
Hello!

I've been studying the Mamba paper and its corresponding implementation in this repository. In the paper, there is a mention of gradient clipping, which is a common technique used to prevent gradient explosion or vanishing. However, upon reviewing the code, I couldn't find the implementation of clip_grad_norm_, which is typically used for gradient clipping in PyTorch.
![image](https://github.com/state-spaces/mamba/assets/60723818/c0aee3ce-1665-4646-a6f8-6a09f813ccb4)

Does the absence of clip_grad_norm_ in the code imply that the Mamba model doesn't encounter issues related to gradient explosion or vanishing? Or is there another mechanism in place to handle these potential problems? I would appreciate any clarification on this matter. 

Thank you.

**Comment by tridao** [comment]
We didn't include training code in this repo, just the model code. We consider gradient clipping to be part of the training code and not part of the model code. Just like we don't include the code for AdamW in this repo even though we use AdamW in our training.

**Comment by EricLina** [comment]
Thanks for you reply, but I am still confused that how does mamba solve gradient explosion or vanishing problem? 
Sovled by gradient clip in training, or by well defined A (Hippo matrix)?

**Comment by tridao** [comment]
Mamba did not aim to solve gradient explosion or vanishing. S4 models already train stably. Gradient clipping is standard in LM training. 
The initialization of A and dt are important, but those were already done in prior work (S4, S4D).

**Comment by EricLina** [comment]
Thanks.
